sympy/physics/control/control_plots.py

```diff
@@ -959,3 +961,187 @@ def bode_plot(system, initial_exp=-5, final_exp=5,
+    k = diff(k_expr, system.var)
+    points = solve(k, system.var)
+    points = [point.evalf() for point in points if point.is_real]
+    breakaway_points = []
+
+    for point in points:
+        if k_expr.subs(system.var, point) > 0:
+            breakaway_points.append(np.array([point + 0j, point + 0j], dtype=np.complex128))
+
+    if (len(breakaway_points)>0): return np.vstack(breakaway_points)
+
+def _find_roots(num, den, k=None):
+    """Private method to find roots efficiently."""
+    if k is not None:
+        num_roots = max(len(np.roots(den)), len(np.roots(num)))
+        roots = []
+
+        for gain in k:
+            individual_roots = np.roots(np.polyadd(den, (gain*num)))
+            if len(individual_roots) == num_roots:
+                individual_roots.sort()
+                roots.append(individual_roots)
+
+        return np.vstack(roots)
+
+    num_roots = np.roots(num)
+    den_roots = np.roots(den)
+
+    return num_roots, den_roots
+
+
+def root_locus_numerical_data(system,  k_max=40, num=100, **kwargs):
+    """
+    Returns the numerical data of Root Locus plot of the system.
+    It is internally used by ``root_locus_plot`` to get the data
+    for plotting Root Locus plot. Users can use this data to further
+    analyse the dynamics of the system or plot using a different
+    backend/plotting-module.
+
+    Parameters
+    ==========
+
+    system : SISOLinearTimeInvariant
+        The system for which the pole-zero data is to be computed.
+    k_max : Number, optional
+        The maximum gain value permitted for root-locus computation.
+        Defaults to 40.
+    num : Number, optional
+        The number of points used for sampling. Defaults to 100.
+
+    Returns
+    =======
+
+    tuple : (real, imag)
+        real = real-axis values of Root Locus plot.
+        imag = imaginary-axis values of Root Locus plot.
+
+    Raises
+    ======
+
+    NotImplementedError
+        When a SISO LTI system is not passed.
+
+        When time delay terms are present in the system.
+
+    ValueError
+        When more than one free symbol is present in the system.
+        The only variable in the transfer function should be
+        the variable of the Laplace transform.
+
+    Examples
+    ========
+
+    >>> from sympy.abc import s
+    >>> from sympy.physics.control.lti import TransferFunction
+    >>> from sympy.physics.control.control_plots import root_locus_numerical_data
+    >>> tf1 = TransferFunction(2*s**2 + 5*s + 1, s**2 + 2*s + 3, s)
+    >>> root_locus_numerical_data(tf1)   # doctest: +SKIP
+    (array([[-1.        , -1.        ],[-1.0000001 , -1.0000001],...,[-2.25458196, -0.23806888],
+    [-2.25880749, -0.23501967]]),array([[-1.41421356,  1.41421356],[-1.41421315,  1.41421315],
+    ...,[ 0.        ,  0.        ],[ 0.        ,  0.        ]]))
+
+    See Also
+    ========
+
+    root_locus_plot
+    """
+    system = system.doit()
+    k = np.logspace(np.log10(1e-6), np.log10(k_max), num=num) - 1e-6
```

Comment:
Hi @oscarbenjamin, I would like if you could go through this once.

>Yes, but we don't need 10000 points to plot the curve. The num=10000 plot above shows that an acceptable plot is obtained from only 40 points (since the other 9960 points are on the real axis).
I expect that uniform spacing is not optimal. Naturally I would guess that something like logarithmic or harmonic spacing would be better. You can see in the num=10000 plot that using uniform intervals for k does not lead to a uniform distance between the points in the plot.
Ideally the curve would be generated using pseudo-arc-length continuation so you can generate them at a fixed 2D distance within the plot but that's probably more complicated than it is worth.

Originally posted by **oscarbenjamin** in this [comment](https://github.com/sympy/sympy/pull/21763#discussion_r675992906) .

I have tried to address this here.
**Consider the earlier attempt at Root Locus Plot**  - c162c48b490166bd82995babadb697b3f7f09823

For `TransferFunction(2*s**2 + 5*s + 1, s**2 + 3*s + 5, s)`
If `num=10000` (sampled 10000 times, default)

![image](https://github.com/sympy/sympy/assets/90216905/634255e9-daf2-49db-addb-e52be808745f)

1. Here `np.linspace` was used due to which majority points are not on curve but on the straight line which is unnecessary. I have used `np.logspace`.
2.  Falsely joined points when curve approaches the axis. I have added the `get_breakaway_points()` function which calculates the points where 2 locus branches meet the axis and added the points to the array at the correct index. Here if you `print(idx)` with `np.linspace ` it gives 3 ( only 3/100 sampled points on the curve ) while with `np.logspace` it gives 78 ( 78/100 points on the curve which is indeed good sampling).
3. Here the issue was, the points were being generated upto `k = 400` but the points were only present between `+- 1.5
` so a `k_max` of 2 would also suffice here for the y axis values . But to get all points on the x-axis you would need a higher value . I have generalized it to 40.

Now we get a better plot with just **100** points compared to the earlier **10000** points required and the time to generate the plot is also lesser.

Reply:
A better approach would be to compute the end points of the lines directly. I don't know much about root locus analysis but as far as I can tell the situation here is something like we have a polynomial and there is a parameter `k` in the coefficients like this:
```
In [1]: k, s = symbols('k, s')

In [2]: p = s**3 + (1 + k)*s**2 + k*s + 1

In [3]: p
Out[3]: 
       3    2            
k⋅s + s  + s ⋅(k + 1) + 1
```
The root locus plot shows how the roots of the polynomial for `s` vary as `k` is changed. For some values of `k` the roots will change from being real to being complex. That happens when the discriminant of the polynomial is zero and we can compute those values of `k` directly:
```
In [4]: d = discriminant(p, s)

In [5]: d
Out[5]: 
 4      3      2           
k  - 6⋅k  + 7⋅k  + 6⋅k - 31

In [6]: [r.n(3) for r in real_roots(d)]
Out[6]: [-1.48, 4.48]
```
We know that the end points of the lines and curves must be at -1.48 and 4.48 so there are 3 sets of k values to consider `k < -1.48`, `-1.48 < k < 4.48` and `4.48 < k`. Picking a rational number in each of these regions allows us to determine how many real roots there are in each case:
```
In [7]: len(real_roots(p.subs(k, -2)))
Out[7]: 3

In [8]: len(real_roots(p.subs(k, 0)))
Out[8]: 1

In [9]: len(real_roots(p.subs(k, 5)))
Out[9]: 3
```
We know now that for `-1.48 < k < 4.48` there are two nonreal roots but for all other values of `k` there are three real roots.

Reply:
Not sure if I totally get what you meant by getting end points by counting the number of `real_roots` in each interval. I tried to prepare a snippet as follows - 

```
import matplotlib.pyplot as plt
from sympy import *
from sympy.physics.control import *
s,k = symbols('s k')
tf = TransferFunction(2*s**2 + 5*s + 1, s**2 + 2*s + 3, s)

eq = tf.den + k * tf.num
d = discriminant(eq, s)
roots_discriminant = [r.n(3) for r in real_roots(d)]
# roots_discriminant = [-0.490, 0.961] - x < -0.490, -0.490< x < 0.961, 0.961< x 
k_values = [-1.5,-1.35,-1.2,-1.05,-0.9,-0.75,-0.6,-0.45,-0.3,-0.15,0,0.15,0.3,0.45,0.6,0.75,0.9,1.05,1.2,1.35,1.5]

real_parts = []
imag_parts = []
for k_value in k_values:
    modified_eq = eq.subs(k, k_value)

    roots = np.roots(modified_eq.as_poly().all_coeffs())
    real_parts.append(np.real(roots))
    imag_parts.append(np.imag(roots))

plt.xlabel('Re')
plt.ylabel('Im')
plt.plot(real_parts, imag_parts)
```

And got a plot - 
![image](https://github.com/sympy/sympy/assets/90216905/bd66ed98-7960-488a-88fe-25927aa1266c)

So even now, it comes down to sampling and finding roots. I guess some kind of sorting will also be needed in each interval and the wrongful joining of points when curve approaches the axis still exists.

Reply:
Can you takeover from here and help me to design a plot. I get how this helps in analysis (presence of complex conjugates or only real roots etc.) but cannot figure out the plot with just endpoints of the intervals
```
print(len(real_roots(eq.subs(k, -2))))
print(len(real_roots(eq.subs(k, 0))))
print(len(real_roots(eq.subs(k, 2))))
2
0
2
```

Reply:
> Can you takeover from here and help me to design a plot.

No, but if you ask a more specific question I can maybe answer it.

> the wrongful joining of points when curve approaches the axis still exists

That is because you are not using the values of k found in roots_discriminant to make the start and end of the lines. Those k values are precisely the ones that correspond to where the curve meets the axis:
```python
In [32]: r1, r2 = real_roots(discriminant(eq, s))

In [34]: roots(eq.subs(k, r1), s)
Out[34]: {5 + √38: 2}

In [35]: roots(eq.subs(k, r2), s)
Out[35]: {5 - √38: 2}

In [36]: N(5 - sqrt(38))
Out[36]: -1.16441400296898
```
As you can see in the plot `-1.16` is where the curves meet. You should use that as the end point of the lines.

Reply:
Once the new plotting module is merged and it is based on the adaptive library, then these plots will lot better at the  locations where derivative of the roots wrt k is steep. If the sympy plotting module is used to construct the plots (instead of numpy and matplotlib directly) then these benefits will be applied once the new plot module is merged.

Reply:
> No, but if you ask a more specific question I can maybe answer it.

Yeah, my apologies. My specific question is once we have calculated the start and end points using discriminant, what role does number of real roots in each interval play in the plot. Could you expand on that ?
As you can see in my snippet above once I get the roots of the discriminant, I limit my k_values around them and eventually do sampling only. I'm not sure I did what was expected correctly.

Reply:
Hi, I could find the boundary points of the curve through this -
```
from sympy import *
s = Symbol('s')
k = Symbol('k')
tf = TransferFunction(2*s**2 +5*s +1, s**2+2*s+3,s)
n = tf.num
d = tf.den
eq = d+k*n
print(eq)                            # k*(2*s**2 + 5*s + 1) + s**2 + 2*s + 3

print(nroots(n))                     # k = oo .Boundary points [-2.28077640640441, -0.219223593595585]

dis = discriminant(eq, s)  
print(d)                             # 17*k**2 - 8*k - 8
d_d = [r.n(3) for r in roots(d)]
print(d_d)                           # [-1.0 - 1.41*I, -1.0 + 1.41*I]   
d_r = [r for r in real_roots(dis)]
print(d_r)                           # [4/17 - 2*sqrt(38)/17, 4/17 + 2*sqrt(38)/17]

print(roots(eq.subs(k, d_r[1]), s))  # {5 - sqrt(38): 2}
print(N(5 - sqrt(38)))               # −1.16441400296898
```

This will gives us a plot like this - 
![image](https://github.com/sympy/sympy/assets/90216905/7ac7511f-194f-44a0-abcf-6dec5f13cce6)

Which is nothing but poles and zeros of `TransferFunction `and point where discriminant is 0. We would need to do the sampling upto that value `k_discriminant` (1.61.. which is quite less than 40) and need considerable points ( less than 100 but still ) for this part - 
`root_locus_plot(tf,k_max= 1.16441400296898,num=50)`
![image](https://github.com/sympy/sympy/assets/90216905/f2c01172-1c60-4465-96c1-ac7f1c1296fa)

Reply:
> Once we have calculated the start and end points using discriminant, what role does number of real roots in each interval play in the plot.

For real roots the plot is just a straight line and so it is not necessary to call real_roots, nroots etc to generate any points for the plot. Just find the coordinates of the start and end of the line and draw the line using those two points.

Reply:
So to be clear, let me point out what I have understood - 
1. Get the endpoints ( the discriminant method is applicable, but these are essentially poles and zeros of the TransferFunction which are directly accessible through `tf.poles()` and `tf.zeros()`) and join them. An issue of coloring will arise here since simply joining them will have the line of the same color. Instead separate branches are supposed to have different colors. I thought If this could be handled using the matplotlib coloring schemes but don't think it can be fixed.
2. Then prepare the curve with better sampling (we don't need a `k_max `value upto 40 to generate the curves). Yet, the `k_max `value will have to be thought about . Also, number of points will be reduced .

@Sc0rpi0n101 I am thinking in the right direction ?

Reply:
> Just find the coordinates of the start and end of the line and draw the line using those two points.

Hi @oscarbenjamin, I have some doubts in the approach you have mentioned.
1. How is the curve supposed to be handled here since this approach talks about the real axis. For that we would anyways need sampling of points to a gain (`k`) value that cannot be generalized. Right now, most of the sampling (90% points) are getting generated on the curve and there are very few points on the real axis, so most plots will need this sampling even if I separately join end points on real axis (there would not be any computational advantage here).
Also, we have to get each point on the real axis and monitor which of them need to be joined (instead of directly joining the end points here)
Consider this example -
![image](https://github.com/sympy/sympy/assets/90216905/bc4c3ca7-c590-4630-9a18-dfd765a6fdfe)

I came across this when I was trying to apply this approach on transfer functions of higher order. Here I was made the parametric equation - found it's roots and played with points in the intervals. According to the count of `real_roots` It suggests there are 4 real roots (I should join -21 and -4) but as you can the MATLAB plot (and plot by this current pr) there should not be a locus on points like -16 and -6.

![image](https://github.com/sympy/sympy/assets/90216905/8c2f3eb7-9701-429c-924b-240709d125b1)

2. Coloring the branches. This can be ignored, but if we want different branches to be colored differently, I'm not sure if we can directly join the end points. I tried a matplotlib coloring scheme and made it work for quadratic transfer functions but I'm not certain it will always work.

I would really appreciate if we can discuss an example end to end before I completely shift to this approach. Thanks.

Reply:
This is the equation (please include copy-pastable code rather than images of code):
```python
In [111]: k = Symbol('k')

In [112]: n = s + 7

In [113]: d = s*(s + 5)*(s + 15)*(s + 20)

In [114]: eq = d + k*n
```
We can find the end points for `k -> 0` and `k -> oo`.
```python
In [115]: real_roots(d)
Out[115]: [-20, -15, -5, 0]

In [116]: real_roots(n)
Out[116]: [-7]
```
There exists a value of `k` between 0 and oo where the discriminant is zero:
```python
In [121]: print(discriminant(eq, s))
-27*k**4 + 47088*k**3 - 37907500*k**2 + 3881250000*k + 1265625000000

In [122]: [r.n(3) for r in real_roots(discriminant(eq, s))]
Out[122]: [-131.0, 308.0]

In [123]: [r.n(3) for r in real_roots(discriminant(eq, s)) if r > 0]
Out[123]: [308.0]
```
We can take that value substitute into `eq` and solve for `s`:
```python
In [138]: [k0] = [r for r in real_roots(discriminant(eq, s)) if r > 0]

In [142]: print(k0)
2*CRootOf(27*x**4 - 23544*x**3 + 9476875*x**2 - 485156250*x - 79101562500, 1)

In [139]: k0.n()
Out[139]: 307.765516989741

In [140]: nroots(eq.subs(k, k0))
Out[140]: 
[-21.8216687209134, -12.5742575561632, -2.80203686146167 + 2.32476007395167e-8⋅ⅈ, -2.80203686146167
 - 2.324760073957e-8⋅ⅈ]
```
Notice that we have a pair of complex roots with small imaginary part. This is a numeric error because the polynomial is ill-conditioned and nroots is not an exact algorithm.

Instead we can substitute the exact value and factorise `eq`. I will put `k0` into an `AlgebraicNumber` for display purposes (it's easier to read then expressions with RootOf):
```python
In [143]: ka = AlgebraicNumber(k0, alias='k0')

In [144]: eq.subs(k, ka)
Out[144]: s⋅(s + 5)⋅(s + 15)⋅(s + 20) + (s + 7)⋅k₀

In [146]: print(eq.subs(k, ka).factor(extension=ka))
(s - 7694401*k0/191756250 - 807*k0**3/26632812500 + 318102*k0**2/6658203125 + 196178/17045)**2*(s**2 + s*(-636204*k0**2/6658203125 + 807*k0**3/13316406250 + 289444/17045 + 7694401*k0/95878125) - 31731*k0**2/30437500 + 6*k0**3/7609375 + 56465/487 + 416926*k0/547875)
```
Notice the squared factor which represents a multiple root. That root for `s` is the `s` value where the lines meet in the plot at this value of `k`:
```python
In [151]: print(eq.subs(k, ka).factor(extension=ka).n())
7.85141057298995*(0.356883242241987*s + 1)**2*(s**2 + 34.3959262770767*s + 274.391282802036)
```
We want the root of the squared factor (use `factor_list` to get the squared factor):
```python
In [154]: roots(0.356883242241987*s + 1)
Out[154]: {-2.80203686146166: 1}
```
As you can see in your plot that is where the curves meet the real axis. Note that this is not far off from what `nroots` returned but they can be very different.

We know now that `k=~308` and `s=~-2.8` corresponds to the significant point in the plot. What we need to know is how many real roots we have on each side:
```python
In [159]: len(real_roots(eq.subs(k, 300)))
Out[159]: 4

In [160]: len(real_roots(eq.subs(k, 310)))
Out[160]: 2
```

Reply:
Resolving as the discussion has been shifted below !